test/integration/filters/test_socket_interface.h

```diff
@@ -0,0 +1,85 @@
+
+#include <functional>
+
+#include "envoy/network/address.h"
+#include "envoy/network/socket.h"
+
+#include "common/network/io_socket_handle_impl.h"
+#include "common/network/socket_interface_impl.h"
+
+#include "absl/synchronization/mutex.h"
+#include "absl/types/optional.h"
+
+/**
+ * TestSocketInterface allows overriding the behavior of the IoHandle interface.
+ */
+namespace Envoy {
+namespace Network {
+
+class TestIoSocketHandle : public IoSocketHandleImpl {
+public:
+  using WritevOverrideType = absl::optional<Api::IoCallUint64Result>(uint32_t index,
+                                                                     const Buffer::RawSlice* slices,
+                                                                     uint64_t num_slice);
+  using WritevOverrideProc = std::function<WritevOverrideType>;
+
+  TestIoSocketHandle(uint32_t index, WritevOverrideProc writev_override_proc,
+                     os_fd_t fd = INVALID_SOCKET, bool socket_v6only = false)
+      : IoSocketHandleImpl(fd, socket_v6only), index_(index),
+        writev_override_(writev_override_proc) {}
+
+private:
+  Api::IoCallUint64Result writev(const Buffer::RawSlice* slices, uint64_t num_slice) override;
+
+  const uint32_t index_;
+  const WritevOverrideProc writev_override_;
+};
+
+/**
+ * TestSocketInterface allows overriding of the behavior of the IoHandle interface of
+ * accepted sockets.
+ * Most integration tests have deterministic order in which Envoy accepts connections.
+ * For example a test with one client connection will result in two accepted sockets. First
+ * is for the client<->Envoy connection and the second is for the Envoy<->upstream connection.
+ */
+
+class TestSocketInterface : public SocketInterfaceImpl {
+public:
+  TestSocketInterface();
+  ~TestSocketInterface() override;
+
+  /**
+   * Override the behavior of the IoSocketHandleImpl::writev() method.
+   * The supplied callback is invoked with the arguments of the writev method.
+   * Returning absl::nullopt from the callback continues normal execution of the
+   * IoSocketHandleImpl::writev() method. Returning a Api::IoCallUint64Result from callback skips
+   * the IoSocketHandleImpl::writev() with the returned result value.
+   * NOTE: this method must be called before any sockets are created to avoid race condition in
+   * setting the `writev_overide_proc_` member.
+   */
+  void overrideWritev(TestIoSocketHandle::WritevOverrideProc writev);
+
+private:
+  // SocketInterface
+  using SocketInterfaceImpl::socket;
+  IoHandlePtr socket(os_fd_t fd) override;
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+  std::string name() const override {
+    return "envoy.extensions.network.socket_interface.test_socket_interface";
+  };
+
+  uint32_t getAcceptedSocketIndex() {
+    absl::MutexLock lock(&mutex_);
+    return accepted_socket_index_++;
+  }
+
+  SocketInterface* const previous_socket_interface_;
+  TestIoSocketHandle::WritevOverrideProc writev_overide_proc_;
```

Comment:
And reason to not just have this be covered by the lock?

Reply:
Made it const and initialized at construction. No need to think about thread safety.